### ISSUE TITLE
New Rule: Class method should not be invoked from an object instance

### ISSUE THREAD

**Issue by Skarvion** [integrated-application-development/sonar-delphi#303]
### Prerequisites

- [X] This rule has not already been suggested.
- [X] This should be a new rule, not an improvement to an existing rule.
- [X] This rule would be generally useful, not specific to my code or setup.

### Suggested rule title

Class method should not be invoked from an object instance

### Rule description

Calling a class method from an object instance would mislead reader that the method is an object method and may affect the object itself. Call class method using class name as prefix.

### Rationale

```delphi
// class declaration and implemeantion
type
  TFoo = class(TObject)
  public
    class procedure Bar;
  end;

implementation

class procedure TFoo.Bar;
begin
  // do something
end;

// calling
var
  FooObj: TFoo;
begin
  FooObj := TFoo.Create;
  FooObj.Bar; // non-compliant

  TFoo.Bar; // compliant
end.
```

**Comment by Cirras** [comment]
Agreed that a rule (or rules) to enforce "class constants/variables/methods should be referenced statically" would be useful.

**Comment by Cirras** [comment]
Class methods can be polymorphic, so we should probably exclude virtual class methods from the rule.
